### PR TITLE
[Docs]: Update gemm_v0 docstring, add dtype coverage tests and API doc

### PR DESCRIPTION
**Pull request by 2Sheldonlaw**
## 文件变更

| 文件 | 类型 | 说明 |
|------|------|------|
| `tilelang/language/ascend.py` | 修改 | 补充 gemm_v0 docstring：A/B 须在 L1、C 须在 L0C，dtype 支持（float16/bfloat16/int8），dtype 组合（float A/B→C=float32，int8 A/B→C=int32），kL0Size 约束（≤4095），init 累加语义 |
| `testing/python/language/test_tilelang_ascend_language_gemm_v0.py` | 修改 | 新增 14 个测试用例（parametrized），覆盖 bfloat16/int8 精度 + dtype 异常边界 + init 累加模式 |
| `docs/api_docs/T.gemm_v0.md` | 新增 | 经验证的 API 文档（修正：A/B 须在 L1 而非 L0A/L0B，新增 kL0Size/n_actual 参数，约束8区分单次 Mmad 调用与整体 tiling 维度，移除 A5 行） |
| `testing/python/language/test_tilelang_ascend_language_elementwise.py` | 修改 | 对既有 elementwise 测试添加 low_priority 标记，减少 PR 阶段 CI 用例数（详见下方"elementwise CI 超时优化"章节） |

---

## gemm_v0 测试用例

| 测试函数 | 用例数 | 测试内容 | 新增覆盖（既有测试未覆盖） |
|---------|--------|--------|---------------------------|
| `test_gemm_v0_bfloat16_precision` | 2 | bfloat16 精度，ascendc + pto | ✅ 新增：既有测试仅覆盖 float16 |
| `test_gemm_v0_int8_transpose_full` | 8 | int8 转置全覆盖（4 种组合 × 2 target） | ✅ 新增：既有测试仅覆盖 2 种转置组合 |
| `test_gemm_v0_dtype_mismatch` | 2 | A/B dtype 不匹配应编译失败 | ✅ 新增：既有测试无异常边界测试 |
| `test_gemm_v0_init_false_accumulation` | 2 | init=False 累加模式验证 | ✅ 新增：既有测试未独立验证 |

所有新增测试均标记 `@pytest.mark.low_priority`，PR 合并时跳过，在 daily/weekly CI 中执行。

## gemm_v0 测试结果

- **14 PASSED**（CI 绿）

## 验证过程中发现的问题

1. **A/B 须在 L1，而非 L0A/L0B**：原 docstring 写"A must be in L0A"，但 L0A 用法编译失败，L1 用法编译+运行+精度均通过。gemm_v0 内部处理 L1→L0A/L0B 搬运
2. **float32 ascendc 和 pto 均不可用**：两个后端编译均失败（详见下方"float32 编译错误"章节）

---

#### 第1层：编译错误 — `Find undefined Variable _`

**现象**：`tilelang.compile` 报 `InternalError: Find undefined Variable _`

**根因**：用户写 `as (cid, _)`，`_` 变量（blockIdx.y）未被 codegen 注册。两层原因叠加：

1. `AscendLowerParallelToVector` pass 的 `is_l0c_input` 检测只检查 `CastNode`，float32（L0C 和 GM 同 dtype，无 Cast）被误判为"不是 L0C 输入"，错误创建 temp UB 中间 buffer
2. `AscendWorkspaceReduction` pass 在处理 workspace copy-back 时，往 IR 注入了对 `_` 变量的引用（`vid_offset = _ * elems / 2`）
3. codegen 遍历到该引用时，因 `_` 未注册而报错

**涉及代码**：
- `src/transform/ascend_lower_parallel_to_vector.cc:612-618`（`is_l0c_input` 检测）
- `src/transform/ascend_workspace_reduction.cc:1194-1204`（vid_offset 注入）
- `src/target/codegen_ascend.cc:762`（`name_hint != "_"` 守卫跳过注册）

#### 第2层：NaN — `Mmad<float,float>` 缺少 HF32 模式

**现象**：编译通过后，运行结果全 NaN

**根因**：910B 的 `Mmad<float, float, float>` 需要先调用 `SetHF32Mode(ENABLE)` 开启 HF32 模式，tilelang 的 `mma` 模板未调用

**涉及代码**：
- `src/tl_templates/ascend/common.h:159`（`Mmad` 调用前缺少 `SetHF32Mode`）
- CANN API：`AscendC::SetHF32Mode`（`asc/include/basic_api/kernel_operator_mm_intf.h`）

#### 第3层：精度全零 — `LoadData2D` float32 分形转换不正确

**现象**：NaN 消除后，结果全零

**根因**：`LoadData2D`（`load_cbuf_to_ca` 硬件指令）在 float32 下的分形转换不正确。float32 的分形尺寸是 16×8（float16 是 16×16），`load_cbuf_to_ca` 可能不正确处理 16×8 分形

**涉及代码**：
- `3rdparty/catlass/include/catlass/gemm/tile/copy_l1_to_l0a.hpp:286-309`（`TileCopyTla` 调用 `LoadData`）
- CANN API：`AscendC::LoadData` → `load_cbuf_to_ca`（`dav_m200/kernel_operator_mm_impl.h`）

### pto 后端：GM↔UB 搬运不支持 float32

**现象**：`tilelang.compile` 报编译错误

**根因**：pto 后端的 `TLoad`（`copy_gm_to_ubuf_align_b32`）和 `TStore`（`copy_ubuf_to_gm_align_b32`）不支持 float32 的目标特性。float32 数据在 GM↔UB 搬运阶段即失败，未进入 Mmad 计算阶段

**涉及代码**：
- `3rdparty/pto-isa/include/pto/npu/a2a3/TLoad.hpp:24`（`copy_gm_to_ubuf_align_b32` 不支持 float32）
- `3rdparty/pto-isa/include/pto/npu/a2a3/TStore.hpp:26`（`copy_ubuf_to_gm_align_b32` 不支持 float32）

**与 ascendc 的区别**：ascendc 的错误发生在 L0C 相关路径（`is_l0c_input` 误判 + `Mmad<float>` 缺 HF32 + `LoadData2D` 分形不正确）；pto 的错误发生更早，在 GM↔UB 数据搬运阶段即失败。两个后端的 float32 不可用是**相互独立的问题**，非同一根因

### ascendc 尝试过的修复方案

| # | 修改文件 | 修改内容 | 解决编译 | 解决 NaN | 解决精度 |
|---|---------|---------|:---:|:---:|:---:|
| 1 | `ascend_lower_parallel_to_vector.cc` | `is_l0c_input` 补充直接 BufferLoad 检测 | ✓ | — | — |
| 2 | `ascend_lower_parallel_to_vector.cc` | `DecomposeExpression` 补充无 Cast L0C→GM 路径 | ✓ | — | — |
| 3 | `common.h` (模板内) | `if constexpr` 加 `SetHF32Mode` | — | ✗ (Bisheng 不编译) | — |
| 4 | `codegen_ascend.cc` | `GemmOpCodegen` 里生成 `SetHF32Mode` C++ 代码 | — | ✓ | ✗ |
| 5 | `codegen_ascend.cc` | 加 `SetHF32TransMode(NEAREST_EVEN)` | — | ✓ | ✗ |
| 6 | `common.h` (模板内) | `kDirectionAlign = true` | — | — | ✗ |
| 7 | `codegen_ascend.cc` + `common.h` | `SetHF32Mode` + `kDirectionAlign` 组合 | — | ✓ | ✗ |
| 8 | `codegen_ascend.cc` + `codegen_ascend_pto.cc` | 去掉 `name_hint != "_"` 守卫 | ✓ | — | — |

**最终决定**：全部撤回，float32 precision 测试不保留（已移除）。ascendc 需 CANN 官方修复 `LoadData2D` / `load_cbuf_to_ca` 对 float32 分形（16×8）的支持；pto 需修复 `TLoad`/`TStore` 对 float32 的 GM↔UB 搬运支持。两个后端修复后另行添加测试。

关联 issue：https://github.com/tile-ai/tilelang-ascend/issues/1016

---

## elementwise CI 超时优化

参考 [tail_block PR](https://github.com/tile-ai/tilelang-ascend/pull/1490) 的 `low_priority` 标签策略，对 `test_tilelang_ascend_language_elementwise.py` 应用 4 条精简规则，减少 PR 阶段冗余用例执行耗时，同时保留全量测试中的完整覆盖。

标签机制详见 [pytest_marker_guide.md](docs/pytest_marker_guide.md)。

#### 1. shape 精简

当多组 shape 配置存在覆盖关系时，保留代表性用例为默认执行，被覆盖的用例标 `low_priority`。

**判定规则**：如果用例 A 的 shape 场景已被用例 B 完整包含，则 A 标为 `low_priority`。

#### 2. reduce kind 精简

`sum` / `max` / `min` 三种 reduce kind 共享相同的底层处理路径，保留 `sum` 为默认执行，`max` / `min` 标 `low_priority`。

#### 3. dtype 精简

`float` (float32) 和 `float16` 保留为默认执行；冷门 dtype（`int16`/`int32`/`uint16`/`uint32`/`int8`/`uint8`/`bfloat16`）标 `low_priority`。

**int-only API 例外**：bitwise 系列和 gatherb 无 float 变体，保留一个代表 dtype（bitwise 保留 `int16`，gatherb 保留 `uint16`），其余冷门 dtype 标 `low_priority`，确保 PR 阶段仍有 smoke 覆盖。

> **与 tail_block PR 的差异**：tail_block 测的是与 dtype 无关的 tail 机制，故 float16 也标 `low_priority`；elementwise 测试本身测 dtype 相关计算，故 float16 保留为默认执行。

#### 4. target 保留

`ascendc` 和 `pto` 两个 target 均保留为默认执行，不做区分。

### 改动总览

| 参数维度 | 改动前 | 改动后 |
|----------|--------|--------|
| `silu` shape (2 组) | 全部默认 | 1 默认 + 1 low_priority |
| `transpose_tiled` shape (6 组) | 全部默认 | 3 默认 + 3 low_priority |
| `transpose_tiled` dtype (4 种) | 全部默认 | float/float16 默认 + int16/int32 low_priority |
| `transpose_block_b16` shape (4 组) | 全部默认 | 2 默认 + 2 low_priority |
| `transpose_block_b32` dtype (3 种) | 全部默认 | float 默认 + uint32/uint16 low_priority |
| `transpose_dtype_target_params` (12 组) | 1 low_priority | float/float16 默认（4 组）+ 其余 8 组 low_priority |
| `add_dtype_target_params` (6 组) | 全部默认 | float/float16 默认（4 组）+ int16/int32 low_priority |
| `pow_dtype_target_params` (5 组) | 全部默认 | float/float16 默认（4 组）+ int32 low_priority |
| `compare_dtype_target_params` out_dtype (2 种) | 全部默认 | int8 默认 + uint8 low_priority |
| `bitwise_and/or/xor/xor_slice/not` dtype (2 种) | 全部默认 | int16 默认 + uint16 low_priority |
| `bitwise_lshift/rshift_slice` dtype (4 种) | 全部默认 | int16/int32 默认 + uint16/uint32 low_priority |
| `vec_max/maxs/min/mins/mul/muls/sub/subs` dtype (4 种) | 全部默认 | float/float16 默认 + int16/int32 low_priority |
| `test_adds` dtype (4 种) | 全部默认 | float/float16 默认 + int16/int32 low_priority |
| `test_relu` dtype (3 种) | 全部默认 | float16/float32 默认 + int32 low_priority |
| `test_cast_scale` dtype (2 种) | 全部默认 | float16 默认 + int32 low_priority |
| `compare_slice/scalar/scalar_slice` out_dtype (2 种) | 全部默认 | int8 默认 + uint8 low_priority |
| `createvecindex_ascendc` dtype (4 种) | 全部默认 | float16/float 默认 + int16/int32 low_priority |
| `createvecindex_pto` dtype (4 种) | 全部默认 | int16 默认 + int32/uint32/uint16 low_priority |
| `gather/gather_slice/gathermask` dtype (7 种) | 全部默认 | float/float16 默认 + 其余 5 种 low_priority |
| `gatherb` dtype (2 种) | 全部默认 | uint16 默认 + uint32 low_priority |
| `test_reduce_max` | 默认执行 | 整函数 low_priority |
| `test_reduce_min` | 默认执行 | 整函数 low_priority |
| `test_reduce_dim0_runtime_smoke` op (3 种) | 全部默认 | sum 默认 + max/min low_priority |
| `test_reduce_clear_false_runtime_merge` op (3 种) | 全部默认 | sum 默认 + max/min low_priority |

### 用例数量统计

| 维度 | 改动前总用例数 | 改动后 PR 执行数 | 改动后全量执行数 |
|------|---------------|-----------------|-----------------|
| elementwise 全文件 | 490 | 276 | 490 |

> **PR 执行数** = 仅无标签用例；**全量执行数** = 无标签 + low_priority 用例（不含 ci_skip）。

PR CI 用例数从 461 降至 276（-40%），全量测试用例数 490 不变。

### 本地验证

```bash
# 查看 PR 场景下实际执行的用例（跳过 low_priority 和 ci_skip）
pytest --collect-only -m "not (low_priority or ci_skip)" \
  testing/python/language/test_tilelang_ascend_language_elementwise.py -q

# 查看被标记为 low_priority 的用例
pytest --collect-only -m "low_priority" \
  testing/python/language/test_tilelang_ascend_language_elementwise.py -q

# 模拟全量测试场景（仅跳过 ci_skip）
pytest -m "not ci_skip" \
  testing/python/language/test_tilelang_ascend_language_elementwise.py -v
```

### CODE REVIEW

File: docs/api_docs/T.gemm_v0.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+### 2.3 Specifications
+
+#### 2.3.1 DataType Support
+
+| Platform | A | B | C |
+|----------|:---:|:---:|:---:|
+| Ascend A2 / A3 | float16, bfloat16, float32, int8 | float16, bfloat16, float32, int8 | float32 (when A/B are floating-point), int32 (when A/B are int8) |
+
+> float32 input: ascendc backend not supported yet (compile error), pto backend available. See [issue #1016](https://github.com/tile-ai/tilelang-ascend/issues/1016)
```
Comment (bingmeiyou) [inline docs/api_docs/T.gemm_v0.md]: issue 不在文档中体现

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    _run_dtype_precision("bfloat16", "float", target)
+
+
+@pytest.mark.low_priority
+@pytest.mark.skipif(
+    not (hasattr(torch, "npu") and torch.npu.is_available()),
+    reason="gemm_v0 correctness requires an Ascend NPU runtime",
+)
+@pytest.mark.xfail(reason="float32 gemm_v0 codegen bug: Find undefined Variable _")
+@pytest.mark.parametrize("target", ["ascendc", "pto"])
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 如已在文档中标明, target可以将ascendc去掉, 并去掉xfail

File: docs/api_docs/T.gemm_v0.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
+# T.gemm_v0
+
+## 1. Description
+
+Performs a fractal matrix multiply-accumulate operation: `C += op(A) × op(B)`, where `op` is an optional transpose. When `init=True`, the accumulator C is cleared before computation, equivalent to `C = op(A) × op(B)`.
```
Comment (bingmeiyou) [inline docs/api_docs/T.gemm_v0.md]: 统一用中文版本

File: docs/api_docs/T.gemm_v0.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+| Parameter | Input/Output | Description | Type | Required/Optional |
+|-----------|-------------|-------------|------|-------------------|
+| A | Input | Left matrix, last two dimensions are matrix dimensions, must be in L1 | tensor | Required |
+| B | Input | Right matrix, last two dimensions are matrix dimensions, must be in L1 | tensor | Required |
+| C | Input/Output | Output matrix (accumulator), last two dimensions are matrix dimensions, must be in L0C | tensor | Required |
+| transpose_A | Input | Whether to transpose matrix A | bool | Optional (default `False`) |
+| transpose_B | Input | Whether to transpose matrix B | bool | Optional (default `False`) |
+| init | Input | Whether to clear accumulator C before computation (`True`: clear then compute C=A×B; `False`: accumulate on existing C, C+=A×B) | bool | Optional (default `False`) |
+| kL0Size | Input | K-axis tile size controlling L1→L0A/L0B data movement granularity, must be a multiple of 16 and ≤ 4095 | int | Optional (default `128`) |
+| n_actual | Input | Runtime output column count (≤ N and multiple of 16), only effective on transpose_B path | int | Optional (default `None`) |
```
Comment (bingmeiyou) [inline docs/api_docs/T.gemm_v0.md]: `n_actual` 参数暂时仅 ascendc 支持，文档需说明

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+# ============================================================
+# Boundary tests
+# ============================================================
+
+
+@pytest.mark.low_priority
+@pytest.mark.parametrize("target", ["ascendc", "pto"])
+def test_gemm_v0_dtype_mismatch(target):
+    """A/B dtype 不一致应编译失败"""
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: """A/B dtype mismatch should fail to compile."""

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
+"""
+T.gemm_v0 supplementary tests: full dtype precision + boundary cases
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 大部分兄弟测试文件都有 `tilelang.cache.clear_cache()` 或 `tilelang.disable_cache()`。缺少它可能导致其他测试的缓存编译结果引起 flakiness。

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+    kernel = _compile(main, target)
+    a = torch.randn(M, K_total, dtype=torch.float16, device="npu")
+    b = torch.randn(K_total, N, dtype=torch.float16, device="npu")
+    c = torch.zeros(M, N, dtype=torch.float16, device="npu")
+    torch.npu.synchronize()
+    kernel(a, b, c)
+    torch.npu.synchronize()
+    ref_c = a @ b
+    torch.testing.assert_close(c, ref_c, rtol=1e-2, atol=1e-2)
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 现有测试文件及大部分兄弟测试文件都以如下代码结尾：
```python
if __name__ == "__main__":
    pytest.main([__file__, "-v", "-n", "8"])
```

新测试文件没有这个块。

**建议**: 添加 `if __name__ == "__main__"` 块以保持一致性。

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
+"""
+T.gemm_v0 supplementary tests: full dtype precision + boundary cases
+
+Covers combinations missing from existing tests:
+1. bfloat16 precision (existing only tests float16)
```
Comment (bingmeiyou) [inline testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 可以将这部分测试用例代码 迁移整合到  test_tilelang_ascend_language_gemm_v0.py 文件中, 不再新增文件

File: docs/api_docs/T.gemm_v0.md (hunk 1 above)
Comment (2Sheldonlaw) [reply docs/api_docs/T.gemm_v0.md]: 已修改

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py (hunk 2 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 已修改

File: docs/api_docs/T.gemm_v0.md (hunk 3 above)
Comment (2Sheldonlaw) [reply docs/api_docs/T.gemm_v0.md]: 已修改

File: docs/api_docs/T.gemm_v0.md (hunk 4 above)
Comment (2Sheldonlaw) [reply docs/api_docs/T.gemm_v0.md]: 已修改

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py (hunk 5 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 已修改

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py (hunk 6 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 已修改

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py (hunk 7 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 已修改

File: testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py (hunk 8 above)
Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_gemm_v0_dtype_coverage.py]: 已将测试迁移到 test_tilelang_ascend_language_gemm_v0.py

File: testing/python/language/test_tilelang_ascend_language_gemm_v0.py [hunk 9]
```diff
@@ -331,5 +331,153 @@ def test_gemm_v0_int8_kl0split_eq1(transpose_A, transpose_B):
...
+    b = torch.randn(full_K, full_N, dtype=torch_dtype, device="npu")
+    c = torch.zeros(full_M, full_N, dtype=torch_dtype, device="npu")
+    torch.npu.synchronize()
+    kernel(a, b, c)
+    torch.npu.synchronize()
+    ref_c = a @ b
+    torch.testing.assert_close(c, ref_c, rtol=1e-2, atol=1e-2)
+
+
+@pytest.mark.low_priority
```
Comment (z00520135) [inline testing/python/language/test_tilelang_ascend_language_gemm_v0.py]: 新增的gemm_v0用例都是低优先级的 评估下是否需要保留些到提交pr时执行

Comment (2Sheldonlaw) [reply testing/python/language/test_tilelang_ascend_language_gemm_v0.py]: 感谢。已保留 7 个核心用例在 PR 执行，7 个仍标 low_priority
本地 7/7 PASSED，详见 9d880a8f。

**Review by z00520135 — APPROVED** [review]

**Review by LLMZhangYC — APPROVED** [review]
approve
